assets/src/block-validation/sidebar.js

```diff
@@ -12,22 +18,24 @@ import { useEffect } from '@wordpress/element';
 import './style.css';
 import AMPValidationErrorsIcon from '../../images/amp-validation-errors.svg';
 import AMPValidationErrorsKeptIcon from '../../images/amp-validation-errors-kept.svg';
+import { Loading } from '../components/loading';
 import { Error } from './error';
 import { BLOCK_VALIDATION_STORE_KEY } from './store';
-import { AMP_VALIDITY_REST_FIELD_NAME } from './constants';
 
 /**
  * Editor sidebar.
  */
 export function Sidebar() {
 	const { setIsShowingReviewed } = useDispatch( BLOCK_VALIDATION_STORE_KEY );
+	const { savePost } = useDispatch( 'core/editor' );
 
-	const { ampCompatibilityBroken, isShowingReviewed, status, reviewLink } = useSelect( ( select ) => ( {
+	const { ampCompatibilityBroken, isDirty, isFetchingErrors, isShowingReviewed, status, reviewLink } = useSelect( ( select ) => ( {
 		ampCompatibilityBroken: select( BLOCK_VALIDATION_STORE_KEY ).getAMPCompatibilityBroken(),
+		isDirty: select( 'core/editor' )?.isEditedPostDirty(),
```

Comment:
I think `isEditedPostDirty` may not be quite the right selector. What we want is whether it is dirty _and_ there have been changes _after_ the last autosave.

Reply:
I've been doing some research around that lately.

It seems that finding out if a dirty post has changes done after the last autosave may not be easy (or rather may be expensive). It's because we have to compare the post content to see if they are different. We can do that in the console like that:

```js
wp.data.select( 'core/editor' ).getEditedPostAttribute( 'content' ) === wp.data.select( 'core/editor' ).getAutosave().content
```

While getting the autosave `content` is cheap (it's stored as a string), getting the currently edited post `content` [may require](https://github.com/WordPress/gutenberg/blob/aaa9e8301ba6c82d707dc402248c928c40f46700/packages/editor/src/store/selectors.js#L996-L1002) the `blocks` serialization.

That's my understanding at this point – I'm planning to double-check it tomorrow. It seems to be the case, though, when considering the [comment](https://github.com/WordPress/gutenberg/blob/aaa9e8301ba6c82d707dc402248c928c40f46700/packages/editor/src/store/selectors.js#L656-L662) in the `isEditedPostAutosaveable()` selector and then expanded in the `hasChangedContent()` selector [comment](https://github.com/WordPress/gutenberg/blob/aaa9e8301ba6c82d707dc402248c928c40f46700/packages/editor/src/store/selectors.js#L109-L114).

If you have any suggestions, please share.

Reply:
I checked the behavior of the Preview button and I can see that it doesn't even try to see if the current state of the editor has changed since the last save/autosave. It just goes ahead and issues the `POST` request to update the autosave revision, and if it hasn't changed then at that point the REST API returns with a 400 error:

```json
{"code":"rest_autosave_no_changes","message":"There is nothing to save. The autosave and the post content are the same.","data":{"status":400}}
```

This doesn't help us, as we want to show a button to initiate an autosave if there have been changes since the last autosave without actually making the request.

Here's an idea: instead of checking if the current editor content is not dirty or it matches the last autosave revision, can we rather store some new `autosaveChanged` boolean state? Consider this:

1. Given a new piece of state called `autosaveChanged` which is initially `false`.
1. Every time the content changes, set `autosaveChanged` to `true`.
1. When doing a full post save (regular revision), set `autosaveChanged` to `true`.
1. After a post autosave, also set `autosaveChanged` to `true`.

In this way we only need to check `autosaveChanged` to see if the “Autosave and re-validate” button should appear. Tapping the button would then cause `autosaveChanged` to be `false` and the button would be hidden.

If `isEditedPostAutosaveable()` is `false`, then none of the autosave logic would be employed and instead just the existing `hasChangedContent()` would be used. For that, the button would be just “Save and re-validate“. Or instead, we could just not have any button at all and the user could click the normal post Update button.

Would that work?

Reply:
An approach similar to the one you suggested here, @westonruter, has been proposed in 08aaa52.

In order to allow revalidating AMP only if something changed since the last autosave, we have to add a custom hook. It tracks the post content changes.

Once the first content change is detected, a `isPostDirty` flag is set in the store and the content is no longer tested (checking post content it's expensive since each time it is requested, the `blocks` list must be serialized).

When the post gets saved or autosaved, the flag is cleared and the content is observed again.

The custom hook is not taking the post meta into consideration yet. This will be fixed next (along with updating the unit tests).